Improve how we build A_and_Aᵀ

amontoison opened the pull request:
close #161 

If we want to store exactly `[0 A'; A 0]`, we can't use `symmetric_pattern` to build `Aᵀ`.
```julia
Aᵀ = if symmetric_pattern || A isa Union{Symmetric,Hermitian}
  A
else
  transpose(A)
end
```
We will have the wrong nonzeros in `A_and_Aᵀ` (but the correct sparsity pattern).

I'm wondering if we should store a different matrix than `b = A_and_Aᵀ` in `StarSetResult` or `TreeSetResult`.  
If I remember correctly, we want to store `B` to have the correct type for the function `decompress`, but in this case, it’s always a `SparseMatrixCSC`, whereas `A` could have a different type.

We probably only want the `AdjacencyGraph` of `A_and_Aᵀ` and not the matrix itself, and we could likely achieve this with a keyword argument for `AdjacencyGraph` or another constructor that takes two arguments (`A` and `Aᵀ`).

Off-topic: We can modify the function  `neighbors(g::AdjacencyGraph, v::Integer)` to not use `filter` in the case of bicoloring because we know that we don't have diagonal coefficients.

gdalle commented:
> We probably only want the AdjacencyGraph of A_and_Aᵀ and not the matrix itself, and we could likely achieve this with a keyword argument for AdjacencyGraph or another constructor that takes two arguments (A and Aᵀ).

You're right, in fact that's how I did it initially in #152 (check out the first commit, which is itself a sub-PR). I removed it afterwards because it felt simpler, but looking at the code again we probably never need to actually materialize `A_and_Aᵀ`. The lazy graph approach would likely work. I have opened #161 to keep track.

gdalle commented on src/interface.jl:
```diff
@@ -232,15 +232,22 @@ function coloring(
+
+    # build A_and_Aᵀ
+    sparse_A = A isa SparseMatrixCSC ? A : SparseMatrixCSC(A)
+    nnzA = nnz(A)
```
```suggestion
    nnzA = nnz(sparse_A)
```
I think that's why the tests fail

gdalle commented on src/interface.jl:
```diff
@@ -232,15 +232,22 @@ function coloring(
+    colptr[1:(n + 1)] .= sparse_A.colptr
+    rowval[1:nnzA] .= sparse_A.rowval .+ n
+    nzval[1:nnzA] .= sparse_A.nzval
+    sparse_Aᵀ = SparseMatrixCSC(sparse_A')
```
```suggestion
    sparse_Aᵀ = SparseMatrixCSC(transpose(sparse_A))
```

amontoison commented:
@gdalle 
I updated this old PR to directly build the augmented adjacency matrix in the format `SparsityPatternCSC`.
I also updated the definition such that `SparsityPatternCSC <: AbstractMatrix{Bool}`.
`StarSetColoringResult` and `TreeSetColoringResult` were not happy that `A_and_Aᵀ` was not an `AbstractMatrix`.

We should check with some benchmarks that we still target the optimized star / acyclic decompression for `SparseMatrixCSC`.

amontoison commented on src/interface.jl:
```diff
@@ -234,18 +234,72 @@ function coloring(
+        end
+
+        colptr[p + 1] = nnzS + counter
+        @assert colptr[p + 1] == 2 * nnzS + 1
```
```suggestion
```

gdalle reviewed (changes requested):
Can we reduce code duplication by using the same code for `transpose` and `bidirectional_pattern`? For instance defining a subroutine like

```julia
transpose_pattern!(A, colptr, rowval)
```

which could then also be applied on `colptr = view(...)` and `rowval = view(...)` in the case of the bigger matrix (and followed by a vectorized addition to account for the shift)

amontoison commented:
> Can we reduce code duplication by using the same code for `transpose` and `bidirectional_pattern`? For instance defining a subroutine like
> 
> ```julia
> transpose_pattern!(A, colptr, rowval)
> ```
> 
> which could then also be applied on `colptr = view(...)` and `rowval = view(...)` in the case of the bigger matrix (and followed by a vectorized addition to account for the shift)

Yes, we could reduce code duplication by implementing a subroutine like `transpose_pattern!`.
However, the gain would only be around 20–30 lines of code (including comments and spacing).

That said, introducing the offset logic for rows, columns, and totals might slightly impact readability.
Personally, I don't think it's worth spending time on this right now.

I suggest we merge the PR as it is.
If you feel strongly about reducing duplication, we could open an issue to revisit this improvement later.

gdalle reviewed (approved):
i don't fully understand the code but I trust the tests I added
